### DISCUSSION TITLE
Run deferrable operators on celery workers

### DISCUSSION THREAD

**Discussion by antoniocorralsierra** [apache/airflow#35373] [Q&A]
Right now I have an Airflow instance working with KubernetesExecutor and I would like to change for CeleryKubernetes Executor. There are a few reason, but one of them is for running the deferrable operators. 

This operators are implemented by me and they are very simply. It only make a request for create the async task, suspend themselves and print the event returned by the Trigger. I try to execute it on celery worker but when I call to defer method the worker die. I know that is possible run these operators with KubernetesExecutor setting queue option to kubernetes (default value) but as I said before, one of the reasons for changes is this.

There is any way to run this type of operators (deferrable) on workers without it died?. For example, suspending the proccess instead of the whole worker.

**Comment by Taragolis** [comment]
[Deferrable operators](https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/deferring.html) run into the separate service which call a [triggerer](https://airflow.apache.org/docs/apache-airflow/stable/cli-and-env-variables-ref.html#triggerer), which should be independent of executor type.

If your worker failed that is a sign that something really go wrong. Maybe deployment, maybe usage or maybe some issue which exists in previous version of airflow and do not persist anymore.

For narrow potential problem could you provide a bit more details, e.g (this list not limited, more details always is better):
- Version of Airflow
- DB Backend Type
- Deployment details
- Logs from worker
- Logs from Airflow Triggerer service
 
And I hope with more details someone could help you

**Reply by Taragolis** [reply]
And have a DAG sample which are reproduce you issue also could be helpful

**Reply by antoniocorralsierra** [reply]
When you have a KubernetesExecutor and run a deferrable operators, one pod is create and it is running until that deferrable operators call to defer method. In this point, the execution is moved to triggerer which is waiting to async task finished. When the async task finish, another pod is up and continue the execution with the method indicate in the trigger parameter "method_name".

One schema of my custom deferrable operator is:

```
class ExampleAsyncTaskOperator(BaseSensorOperator):

   def __init__():
       #TODO
  
   def execute(self, context):
       code doing some checks (if, else)
       send_request ( send request to external service that create an async task)
       # Until this point, all operations run in a pod that create kuberneteExecutor. This pod die when defer is called.
       self.defer(trigger=custom_trigger, method_name="execute_complete", timeout=100) # Here the triggerer action begin

 def execute_complete(self, context, event=None):
     # This function also run in a new pod not in the triggerer  
      print_event
```

If you run with CeleryExecutor, the method "execute" run in the celery worker and when you call to defer method the worker die.

**Reply by Taragolis** [reply]
That is not help to figure out what is the root of the problem in your situation.

Did [example_time_delta_sensor_async](https://github.com/apache/airflow/blob/main/airflow/example_dags/example_time_delta_sensor_async.py) works fine in your cluster?

**Reply by antoniocorralsierra** [reply]
Solved. It was not a problem due to deferrable operators, it was that the worker pod didn't have memory enough. The pod always died when was running deferrable operator so I thought that defer method was the motive. But after read the worker log I saw the following message: worker: warm shutdown (mainprocess). Increasing the resouces assigned to worker it works fine.

Thanks for the help :)
